### PR TITLE
add a-tiny-js-world-2 task

### PR DESCRIPTION
**Pull request by allykrupina**
_Complete the form below and remove this line._

### A-tiny-js-world-2 

[Demo](https://allykrupina.github.io/a-tiny-js-world/) |
[Code base](https://github.com/allykrupina/allykrupina.github.io/tree/master/a-tiny-js-world)

The code is submitted in a dedicated feature branch.
Please, review.

### CODE REVIEW

File: submissions/allykrupina/a-tiny-js-world-2/index.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+// ======== OBJECTS DEFINITIONS ========
+
+class Inhabitant {
+    constructor(species, name, gender, legs, hands, saying) {
+        this.species = species;
+        this.name = name;
+        this.gender = gender;
+        this.legs = legs;
+        this.hands = hands;
```
Comment (OleksiyRudenko) [inline submissions/allykrupina/a-tiny-js-world-2/index.js]: A base class should contain only properties common for all children.
Hands property obviously doesn't qualify.
Instead a human has to have a unique property.
Please, fix.

File: submissions/allykrupina/a-tiny-js-world-2/index.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        this.saying = saying;
+    }
+}
+
+class Woman extends Inhabitant {
+    constructor(species, name, gender, saying, legs, hands) {
+        super(...arguments);
```
Comment (OleksiyRudenko) [inline submissions/allykrupina/a-tiny-js-world-2/index.js]: We so not need to ask this very concrete class user to specify neither species nor gender.
You can specify this data explicitly when calling `super`.

Please, fix.

Same is applicable to `Man` and also, to a certain extent, to `Dog` and `Cat`.

File: submissions/allykrupina/a-tiny-js-world-2/index.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+class Man extends Inhabitant {
+    constructor(species, name, gender, saying, legs, hands) {
+        super(...arguments);
+    }
+}
+
+class Dog extends Inhabitant {
+    constructor(species, name, gender, saying, legs, hands) {
+        super(...arguments);
```
Comment (OleksiyRudenko) [inline submissions/allykrupina/a-tiny-js-world-2/index.js]: Asking this class user to specify `hands` looks confusing.
Please, don't.

Also applicable to `Cat`.

File: submissions/allykrupina/a-tiny-js-world-2/index.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+const inhabitants = [
+    new Woman('human', 'Nika', 'female', 'I love my cat', 2, 2),
+    new Man('human', 'Kurt', 'male', 'I love my dog', 2, 2),
+    new Dog('dog', 'Grey', 'male', 'wooooof', 4, 0),
+    new Cat('cat', 'Assol', 'female', 'murrr', 4, 0)
+];
+
+function inhabitantToString(obj){
+    return ['<em>' + obj.species + '</em>', '<strong>' + obj.name + '</strong>', obj.gender, obj.legs, obj.hands, obj.saying].join ('; ');
+}
```
Comment (OleksiyRudenko) [inline submissions/allykrupina/a-tiny-js-world-2/index.js]: Please, make this a method of the basemost class.
You may also want to overload this method for humans supplementing a call to parent class method (e.g. `super.toString()`) with unique info re hands.

**Comment by allykrupina** [comment]
@OleksiyRudenko Updated the code according to changes request

**Comment by allykrupina** [comment]
@OleksiyRudenko One more try) Made changes according to Your request.

File: submissions/allykrupina/a-tiny-js-world-2/index.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+class Woman extends Human {
+    constructor(species, name, gender, saying) {
+        super(...arguments);
+    }
+}
+
+class Man extends Human {
+    constructor(species, name, gender, saying) {
+        super(...arguments);
```
Comment (OleksiyRudenko) [inline submissions/allykrupina/a-tiny-js-world-2/index.js]: A quotation from the past comment. Please, pay special attention to what is in **bold* font here.

> We so not need to ask this very concrete class user to specify neither species nor gender.
> You can specify this data explicitly when calling super.

> Please, fix.

> **Same is applicable to Man** and also, to a certain extent, to Dog and Cat.

File: submissions/allykrupina/a-tiny-js-world-2/index.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+class Inhabitant {
+    constructor(species, name, gender, saying) {
+        this.species = species;
+        this.name = name;
+        this.gender = gender;
+        this.saying = saying;
+    }
+    toString(){
+        return ['<em>' + this.species + '</em>', '<strong>' + this.name + '</strong>', this.gender, this.saying, this.legs + '; '].join ('; ');
+    }
```
Comment (OleksiyRudenko) [inline submissions/allykrupina/a-tiny-js-world-2/index.js]: If you use some property in a base class it is an indication that you have to decide:
1) probably this is some common property that also has to belong to this base class, and consequently get initialized here, expected as an argument in `constructor()`, being passed from child classes via `super` etc
or
2) it is not common for child classes and therefore should not be used in a base class, but in some of concrete child classes instead

Please, decide and make `Inhabitant` consistent.

As for the `legs` I would vote for option 1.

File: submissions/allykrupina/a-tiny-js-world-2/index.js [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+        this.saying = 'wooooof';
+    }
+}
+
+class Cat extends Inhabitant {
+    constructor(name, gender, saying) {
+        super('cat', name, gender, saying);
+        this.legs = 4;
+        this.saying = 'murrr';
+    }
```
Comment (OleksiyRudenko) [inline submissions/allykrupina/a-tiny-js-world-2/index.js]: Why not pass `'murrr'` as an argument to `super`?
If you want saying to be `'murr'` disregarding what user passes at construction time, why asking for a saying at all?

**Comment by OleksiyRudenko** [comment]
@allykrupina Please, also resolve conversations that you believe you have properly elaborated on.

File: submissions/allykrupina/a-tiny-js-world-2/index.js (hunk 7 above)
Comment (allykrupina) [reply submissions/allykrupina/a-tiny-js-world-2/index.js]: Fixed

File: submissions/allykrupina/a-tiny-js-world-2/index.js (hunk 6 above)
Comment (allykrupina) [reply submissions/allykrupina/a-tiny-js-world-2/index.js]: Done, if I understood correct.

**Comment by OleksiyRudenko** [comment]
@allykrupina Please, ask questions if anything is not clear. Also try to explain your decisions. This would be handy to render a help based on your individual way of thinking.

**Comment by allykrupina** [comment]
@OleksiyRudenko it seems I started understanding how it works) I checked myself against all past comments. I also updated demo according to last commit.

File: submissions/allykrupina/a-tiny-js-world-2/index.js (hunk 7 above)
Comment (OleksiyRudenko) [reply submissions/allykrupina/a-tiny-js-world-2/index.js]: Please, explain. Cannot how did you implement the suggestion to use a literal string.

File: submissions/allykrupina/a-tiny-js-world-2/index.js [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+class Woman extends Human {
+    constructor(name, gender, saying) {
+        super(name, gender, saying);
+    }
+}
+
+class Man extends Human {
+    constructor(name, gender, saying) {
+        super(name, gender, saying);
```
Comment (OleksiyRudenko) [inline submissions/allykrupina/a-tiny-js-world-2/index.js]: Do we need to ask a user for gender when instantiating `Man`?
Same question re `Woman`.

Comment (OleksiyRudenko) [reply submissions/allykrupina/a-tiny-js-world-2/index.js]: > this.saying = 'murrr';
> This string I already deleted. I call cat/dog saying in the same manner as man/women saying. Is it correct?

You do not need to set `this.saying` as `super` can do it for you. Just pass it a literal.

Comment (OleksiyRudenko) [reply submissions/allykrupina/a-tiny-js-world-2/index.js]: > For second comment. Logically. I didn't think about this).
> Can I delete classes Woman and Man? And call one class Human with gender parameter?
Yes, you can do this. No problem with having `Man` and `Woman` classes. Removal will just make the code smaller so navigation across will become easier, yet the hierarchy will become more consistent. Anyway, you do not have `SheDog` and `HeDog` ;)

Comment (allykrupina) [reply submissions/allykrupina/a-tiny-js-world-2/index.js]: for example
1.without man/woman class
image

or

2. gender in super();
image

Can I add saying to super()  like this?
image

Comment (OleksiyRudenko) [reply submissions/allykrupina/a-tiny-js-world-2/index.js]: Just a note: images are not the best code presenters. Textual code enables copy-pasting, e.g. to give a quick example based on real suggested code.

Markdown offers nice code formatting tool.
Start with `` ```javascript`` (or just `` ``` ``)
Then put your code.
Then end with `` ``` ``

How it looks when rendered
```javascript
class Inhabitant {
     constructor(species, name, gender, saying, legs = 4) {
         this.species = species;
         this.name = name;
         this.gender = gender;
         this.saying = saying;
         this.legs = legs;
     }
```

Another tool is suggestions in code review mode:
image
1. Move the mouse pointer onto space between row number and "+" symbol (at a row you want to start a suggestion for), and click blue plus button that will appear.
1. Click the **Insert a suggestion** button (see black tooltip on the screenshot above)
1. Replace template text with your code.

Try it. There is always a cancel button.

Comments on your suggested code comes next.

File: submissions/allykrupina/a-tiny-js-world-2/index.js [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    }
+}
+
+class Dog extends Inhabitant {
+    constructor(name, gender, saying) {
+        super('dog', name, gender, saying);
+    }
```
Comment (OleksiyRudenko) [inline submissions/allykrupina/a-tiny-js-world-2/index.js]: @allykrupina Here is how a github suggestion looks:
```suggestion
class Dog extends Inhabitant {
     constructor(name, gender) {
         super('dog', name, gender, 'woof!');
     }
```

File: submissions/allykrupina/a-tiny-js-world-2/index.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+        return ['<em>' + this.species + '</em>', '<strong>' + this.name + '</strong>', this.gender, this.saying, this.legs + '; '].join ('; ');
+    }
+}
+
+class Human extends Inhabitant {
+    constructor(species, name, gender, saying, legs = 2, hands = 2) {
+        super(species, name, gender, saying);
+        this.hands = hands;
+        this.legs = legs;
+    }
```
Comment (OleksiyRudenko) [inline submissions/allykrupina/a-tiny-js-world-2/index.js]: Important note `super()` resolves into `ExtendedClass.constructor()`.
So in class `Human` which extends `Inhabitant` a call to `super()` resolves into `Inhabitant.constructor()`.

**A side note:** Why not to call `SomeParentClass.constructor()` explicitly? For you could insert a class in between and wouldn't have to rename `SomeParentClass.whatever`

Ok. So here `super` resolves to `Inhabitant.constructor()`, and `Inhabitant.constructor()` expects params: `species, name, gender, saying, legs`. You do not supply `legs` to `super`. But instead do `this.legs = legs;`. Is there a good reason for that?

Comment (allykrupina) [reply submissions/allykrupina/a-tiny-js-world-2/index.js]: @OleksiyRudenko I corrected legs and super. 
But didn`t understand a little bit about SomeParentClass.constructor().
Do I have to create one more intermediate class?

Comment (OleksiyRudenko) [reply submissions/allykrupina/a-tiny-js-world-2/index.js]: @allykrupina No, you don't have to add yet another class.
That was an attempt to explain what `super` actually is.

E.g. we have following code:
```
class Citizen {
  constructor(fisrtName, lastName, citizenship) {
    this.firstName = firstName;
    this.lastName = lastName;
    this.citizenship = citizenship;
  }
}

class Ukrainian extends Citizen {
  constructor(fisrtName, lastName, gender) {
    super(fisrtName, lastName, 'Ukraine'); // this is actually a call of `Citizen.constructor()`
    this.gender = gender;
  }
}
```

Does example above explain?

Comment (allykrupina) [reply submissions/allykrupina/a-tiny-js-world-2/index.js]: @OleksiyRudenko yes, I understood the example. And as for this.legs = legs; I supplied it to super now, there is no such need to separate it.

**Pull request merged**